iotile_analytics_interactive/iotile_analytics/interactive/reports/cloud_uploader.py

```diff
@@ -0,0 +1,232 @@
+                the slug of the organization that it should be uploaded for.
+
+
+        Returns:
+            str: A URL where the uploaded report can be accessed.
+        """
+
+        if group is not None:
+            slug = group.source_info.get('slug')
+            org = group.source_info.get('org')
+
+            if slug is None or org is None:
+                raise ArgumentError("The group provided did not have org or source slug information in its source_info", slug=slug, org=org)
+        elif org is None or slug is None:
+            raise UsageError("If you do not specify an AnalysisGroup you must pass an org and a slug to this function", org=org, slug=slug)
+
+        report_id = self._create_report(label, slug, org)
+        self._logger.debug("Created report id: %s", report_id)
+
+        keys = self._clean_file_paths(files)
+
+        urls, fields = self._get_signed_urls(report_id, keys)
+        self._upload_files_to_s3(urls, fields, keys, files)
+        self._notify_upload_success(report_id, keys[0])
+
+    def _clean_file_paths(self, files):
+        """Return a suitable s3 keys for all files."""
+
+        first = files[0]
+        basedir = os.path.dirname(first)
+        if basedir is not None:
+            keys = [os.path.relpath(x, basedir) for x in files]
+        else:
+            keys = [x for x in files]
+
+        keys = [x.replace('\\', '/') for x in keys]
+        return keys
+
+    def _create_report(self, label, slug, org):
+        payload = {
+            'org': org,
+            'label': label,
+            'source_ref': slug
+        }
+
+        resp = self._api.report.generated.post(payload)
+        self._logger.debug("Create generated report record api response: %s", resp)
+
+        report_id = resp.get('id')
+        if report_id is None:
+            raise CloudError("No report id returned by cloud", response=resp)
+
+        return report_id
+
+    def _get_signed_urls(self, report_id, files):
+        url_infos = [self._get_signature_url_payload(report_id, x, i > 0) for i, x in enumerate(files)]
+
+        urls = [x[0] for x in url_infos]
+        payloads = [x[1] for x in url_infos]
+
+        resp_list = self._session.post_multiple(urls, payloads, message="Getting file upload authorization", include_auth=True)
+        decoded_resps = [json.loads(resp.decode('utf-8')) for resp in resp_list]
+        urls = [x['url'] for x in decoded_resps]
+        fields = [x['fields'] for x in decoded_resps]
+        return urls, fields
+
+    def _get_signature_url_payload(self, report_id, file_name, public=True):
+
+        payload = {
+            'name': file_name,
+            'acl': 'public-read' if public is True else 'private',
+            'content_type': self._get_mime_type(file_name)
+        }
+
+        resource = self._api.report.generated(report_id).uploadurl
+
+        return resource.url(), payload
+
+    @classmethod
+    def _get_mime_type(cls, filename):
```

Comment:
Just FYI: The cloud is now doing this exact same thing if you don't pass the content_type